File: src/main/scala/net/sourceforge/reb4j/scala/charclass/PredefinedClass.scala
package net.sourceforge.reb4j.scala.charclass

/**
 * Class representing a predefined character class.
 */
class PredefinedClass private[charclass] (val nameChar : Char) 
	extends CharClass
	with SelfContained
	with Union.Subset
	with Intersection.Superset
{
	protected final def invertedNameChar = 
		if (nameChar.isUpper) 
			nameChar.toLower
		else
			nameChar.toUpper
	override def ^ = new PredefinedClass(invertedNameChar)
	override def unitableForm = "\\" + nameChar
	override def equals(other : Any) = other match
	{
		case that : NamedPredefinedClass => false
		case that : PredefinedClass => this.nameChar == that.nameChar
		case _ => false
	}
	override lazy val hashCode = 31 * nameChar.hashCode
}


/**
 * Class representing a predefined character class referenced by name.
 */
final class NamedPredefinedClass private[charclass] (
		nameChar : Char, 
		val className : String
	) extends PredefinedClass(nameChar)
{
	def this (className : String) = this('p', className)
	override def ^ = new NamedPredefinedClass(invertedNameChar, className)
	override lazy val unitableForm = super.unitableForm + "{" + className + "}"
	override def equals(other : Any) = other match
	{
		case that : NamedPredefinedClass =>
			this.nameChar == that.nameChar &&
			this.className == that.className
		case _ => false
	}
	override lazy val hashCode = 
		31 * nameChar.hashCode + className.hashCode
}

File: net.sourceforge.reb4j/src/main/java/net/sourceforge/reb4j/charclass/PredefinedClass.java
package net.sourceforge.reb4j.charclass;

import fj.data.LazyString;

public class PredefinedClass extends AbstractUnitableIntersectable
{
	private static final long serialVersionUID = 1L;
	public final char nameChar;
	
	protected final char invertedNameChar()
	{
		return Character.isUpperCase(nameChar) ? 
			Character.toLowerCase(nameChar) : Character.toUpperCase(nameChar); 
	}
	
	PredefinedClass(final char nameChar)
	{
		this.nameChar = nameChar;
	}

	@Override
	public PredefinedClass negated()
	{
		return new PredefinedClass(invertedNameChar());
	}

	@Override
	public LazyString unitableForm()
	{
		return LazyString.str("\\").append(Character.toString(nameChar));
	}

	@Override
	public final LazyString independentForm()
	{return unitableForm();}

	@Override
	public int hashCode()
	{
		final int prime = 31;
		int result = 1;
		result = prime * result + nameChar;
		return result;
	}

	@Override
	public boolean equals(final Object obj)
	{
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		final PredefinedClass other = (PredefinedClass) obj;
		return nameChar == other.nameChar;
	}

}
